Add KIC 2.1 documentation (#3518)
Kong/docs.konghq.com: app/kubernetes-ingress-controller/2.1.x/guides/upgrade.md

---
title: Upgrading to 2.1.x
---

This guide walks through backwards incompatible changes in the Kong Kubernetes

> Ingress Controller (KIC) from to v2.1.x to help operators evaluate if any

Reviewer: ```suggestion
Ingress Controller (KIC) to v2.1.x to help operators evaluate if any
```

> **Note:** Deploying and upgrading via Helm is the supported mechanism for
production deployments of KIC. If you're deploying KIC using Kustomize or
some other mechanism, you need to develop and test your own upgrade strategy
based on the following examples.

[helm]:https://helm.sh/
[chart]:https://github.com/kong/charts
[list-releases]:https://v3.helm.sh/docs/helm/helm_list/
[helm-docs]:https://v3.helm.sh/docs/

## Upgrading from 2.0.x to 2.1.x

This document primarily covers the requirements for upgrading from 1.x to 2.x.
If you have already upgraded to 2.0.x, only the steps in this subsection are
necessary. If you are still on 1.x, follow the subsequent sections to upgrade
to 2.0.x first. **You must upgrade to 2.0.x before upgrading to 2.1.x.**

### Update CRDs

The previous KongIngress CRD included several incorrectly-named fields which
did not apply correctly to Kong upstreams. 2.1.x CRDs fix these fields:

* `healthchecks.passive.unhealthy.timeout` is now
  `healthchecks.passive.unhealthy.timeouts`
* `healthchecks.active.unhealthy.timeout` is now
  `healthchecks.active.unhealthy.timeouts`

> Review your KongIngresses to see if you use either of the old fields. You will
> need to re-add them with the new field name after upgrading. Kubernetes does
> not allow unknown fields in CRDs, and will strip the old field and its value
> after updating. There is no means of copying these into the new field
> automatically.

Reviewer: ```suggestion
Before upgrading, review your KongIngresses to see if you use either of the old 
fields. You will need to manually re-add them with the new field name after 
upgrading. 

Kubernetes does not allow unknown fields in CRDs. Since the old fields don't 
exist in 2.1, Kubernetes will strip the old field and its value after updating. 
There is no automated way to copy the old values into the new field.
```

```shell
$ helm upgrade ${YOUR_RELEASE_NAME} kong/kong \
  --namespace ${YOUR_NAMESPACE} \
  -f ${PATH_TO_YOUR_VALUES_FILE}
```

After the upgrade completes there is a brief period of time before the new
resources are online. You can wait for the relevant Pod resources to complete
by watching them in your release namespace:

```shell
$ kubectl -n ${YOUR_RELEASE_NAMESPACE} get pods -w
```

Once the new pods are in a `Ready` state, the upgrade is complete.

### Rollback

If you run into problems during or after the upgrade, Helm provides a
rollback mechanism to revert to a previous revision of the release:

```shell
$ helm rollback --namespace ${YOUR_RELEASE_NAMESPACE} ${YOUR_RELEASE_NAME}
```

You can wait for the rollback to complete by watching the relevant Pod
resources:

```shell
$ kubectl -n ${YOUR_RELEASE_NAMESPACE} get pods -w
```

After a rollback, if you ran into issues in production,
consider using a [testing environment](#testing-environment) to
identify and correct these issues, or reference the
[troubleshooting documentation][troubleshooting].

[troubleshooting]:/kubernetes-ingress-controller/{{page.kong_version}}/troubleshooting/

> [admission]:/kubernetes-ingress-controller/{{page.kong_version}}/deployment/admission-webhook.md

Reviewer: ```suggestion
[admission]:/kubernetes-ingress-controller/{{page.kong_version}}/deployment/admission-webhook
```
